File: code/typescript-for-csharp-java/compare/L02Nullable.java
// compare/L02Nullable.java
import java.util.Map;

public class L02Nullable {
    static String initial(String name) {
        return name.substring(0, 1); // nothing in the type says name can be null
    }

    public static void main(String[] args) {
        var tunings = Map.of("standard", "EADGBE");
        String dropD = tunings.get("drop D");
        try {
            System.out.println(initial(dropD));
        } catch (NullPointerException e) {
            System.out.println("initial(dropD): " + e.getClass().getSimpleName());
        }
    }
}

File: code/typescript-for-csharp-java/compare/l02_nullable.cs
// compare/l02_nullable.cs
using System.Collections.Generic;

string Initial(string? name) => name.Substring(0, 1); // CS8602, a warning: the program still builds

var tunings = new Dictionary<string, string> { ["standard"] = "EADGBE" };
tunings.TryGetValue("drop D", out var dropD);
string certain = dropD!; // the same ! as TypeScript, with the same meaning: trust me

try { Console.WriteLine(Initial(null)); }
catch (NullReferenceException e) { Console.WriteLine($"Initial(null): {e.GetType().Name}"); }
try { Console.WriteLine(certain.Length); }
catch (NullReferenceException e) { Console.WriteLine($"certain.Length: {e.GetType().Name}"); }
